src/panels/lovelace/cards/hui-shopping-list-card.ts

```diff
@@ -0,0 +1,217 @@
+    this.config = config;
+    this.items = [];
+  }
+
+  protected render() {
+    if (!this.config || !this.hass) {
+      return html``;
+    }
+
+    // TODO Localize 'Add Item' text
+    return html`
+      ${this.renderStyle()}
+      <ha-card .header="${this.config.title}">
+        <div class="addRow">
+          <paper-input
+            id='addBox'
+            placeholder='Add Item'
+            @keydown='${(e) => this._addKeyPress(e)}'
+            no-label-float
+          ></paper-input>
+          <paper-icon-button
+            slot="item-icon"
+            icon="hass:plus"
+            @click='${(e) => this._addItem(e)}'
+          ></paper-icon-button>
+          <paper-icon-button
+            slot="item-icon"
+            icon="mdi:notification-clear-all"
+            @click='${(e) => this._clearCompleted()}'
+          ></paper-icon-button>
+        </div>
+
+      ${Object.values(this.items).map(item =>
+        html`
+        <paper-icon-item>
+          <paper-checkbox
+            slot="item-icon"
+            ?checked=${item.complete}
+            @click='${(e) => this._itemCompleteTapped(e, item.id)}'
+            tabindex='0'
+          ></paper-checkbox>
+          <paper-item-body>
+            <paper-input
+              id='editBox'
+              no-label-float
+              value='${item.name}'
+              @change='${(e) => this._saveEdit(e, item.id)}'
+            ></paper-input>
+          </paper-item-body>
+        </paper-icon-item>
+        `)}
+      </ha-card>
+    `;
+  }
+
+  private renderStyle() {
+    return html`
+      <style>
+        .addRow {
+          padding-left: 16px;
+          display: inline-block;
+        }
+        .addRow > paper-icon-button {
+          display: inline-block;
+        }
+        .addRow > paper-input {
+          display: inline-block;
+        }
+        paper-icon-item {
+          border-top: 1px solid var(--divider-color);
+        }
+        paper-icon-item:first-child {
+          border-top: 0;
+        }
+        paper-checkbox {
+          padding: 11px;
+        }
+        paper-input {
+          --paper-input-container-underline: {
+            display: none;
+          }
+          --paper-input-container-underline-focus: {
+            display: none;
+          }
+          position: relative;
+          top: 1px;
+        }
+      </style>
+    `;
+  }
+
+  private connectedCallback() {
+    super.connectedCallback();
+    this._fetchData = this._fetchData.bind(this);
+
+    this.hass.connection
+      .subscribeEvents(this._fetchData, "shopping_list_updated")
+      .then(
+        function(unsub) {
+          this._unsubEvents = unsub;
+        }.bind(this)
+      );
+    this._fetchData();
+  }
+
+  private disconnectedCallback() {
+    super.disconnectedCallback();
+    if (this._unsubEvents) {
+      this._unsubEvents();
+    }
+  }
+
+  private _fetchData() {
+    this.hass.callApi("GET", "shopping_list").then(
```

Comment:
We can't just start copy/pasting all these methods from file to file… that's unmaintainable.

Create a new folder `src/data`. In that folder we're going to create a new folder for each component. So create `src/data/shopping_list/data.js` and add 
```js
export fetchItems = (hass) => hass.callApi("GET", "shopping_list");
```

Do not include the `reverse` call in `data.js`, that's an implementation detail of this card.

Do this for all API calls.

Reply:
does it make sense to create a folder for each component to house just a single file in it?

why not just `src/data/shopping_list.js` or `src/data/components/shopping_list.js`?